backend/v3/domain/invoker_event.go

```diff
@@ -0,0 +1,81 @@
+package domain
+
+import (
+	"context"
+
+	"github.com/zitadel/zitadel/backend/v3/storage/eventstore"
+	legacy_es "github.com/zitadel/zitadel/internal/eventstore"
+)
+
+// eventStoreInvoker checks if the [Commander].Events function returns any events.
+// If it does, it collects the events and publishes them to the event store.
+type eventStoreInvoker struct {
+	collector *eventCollector
+}
+
+func newEventStoreInvoker(next Invoker) *eventStoreInvoker {
+	return &eventStoreInvoker{collector: &eventCollector{next: next}}
+}
+
+type EventProducer interface {
+	// Events returns the events that should be pushed to the event store after the command is executed.
+	// If the command does not produce events, it should return nil or an empty slice.
+	Events(ctx context.Context, opts *InvokeOpts) ([]legacy_es.Command, error)
```

Comment:
I've been reading here https://google.github.io/styleguide/go/decisions#nil-slices and on Reddit. General consensus is to always return `nil` if the slice is empty. So I would update the comment.